### PR TITLE
[SofaMiscFEM_test] Add class to test and compare TriangleFEM and TriangularFEMForceField

### PR DESCRIPTION
**Pull request by epernod**
This PR add a new class to test at the same time TriangleFEMForceField and TriangularFEMForceField. Same results and value are expected in the idea of merging those 2 components in a near future. 

In this PR 14 tests are added (7 on each version). 2 are failing and has been disabled: 
[  FAILED  ] TriangleFEMForceField3_test.checkTriangularFEMForceField_init -> inconsistency  in code
[  FAILED  ] TriangleFEMForceField3_test.checkTriangularFEMForceField_values -> inconsistency  in code

This PR also fix the init of both component, checking the input Data value and setting them as required.
Update setter as well which should be upgraded to Data callback in a next PR

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: modules/SofaMiscFem/SofaMiscFem_test/TriangleFEMForceField_test.cpp [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+    }
+
+
+
+
+    void checkCreation(int FEMType)
+    {
+        createSingleTriangleFEMScene(FEMType, 100, 0.3, "large");
+
+        MState::SPtr dofs = m_root->getTreeObject<MState>();
```
Comment (fredroy) [inline modules/SofaMiscFem/SofaMiscFem_test/TriangleFEMForceField_test.cpp]:
```suggestion
        auto dofs = m_root->getTreeObject<MState>();
```

**Review by fredroy — CHANGES_REQUESTED** [review]
to fix the compilation error with gcc/clang

**Comment by fredroy** [comment]
[ci-build]

**Comment by guparan** [comment]
Nice work @epernod :+1:
If you already know what must be done to fix the tests, I'm for doing it in this PR !

**Comment by epernod** [comment]
> 
> 
> Nice work @epernod 👍
> If you already know what must be done to fix the tests, I'm for doing it in this PR !

ok I fixed some of them. The last 2 failing tests need big changes. we can comment them for the moment.

File: modules/SofaMiscFem/src/SofaMiscFem/TriangleFEMForceField.inl [hunk 2]
```diff
@@ -123,9 +120,13 @@ void TriangleFEMForceField<DataTypes>::init()
     _strainDisplacements.resize(_indexedElements->size());
     _rotations.resize(_indexedElements->size());
 
+    if (method == SMALL) {
+        initSmall();
+    }
+    else {
+        initLarge();
+    }
```
Comment (guparan) [inline modules/SofaMiscFem/src/SofaMiscFem/TriangleFEMForceField.inl]:
```suggestion
    if (method == SMALL)
    {
        initSmall();
    }
    else
    {
        initLarge();
    }
```

File: modules/SofaMiscFem/src/SofaMiscFem/TriangleFEMForceField.inl [hunk 3]
```diff
@@ -139,9 +140,13 @@ void TriangleFEMForceField<DataTypes>::reinit()
     else if (f_method.getValue() == "large")
         method = LARGE;
 
+    if (method == SMALL) {
+        //    initSmall();  // useful ? The rotations are recomputed later
+    }
+    else {
+        initLarge(); // compute the per-element strain-displacement matrices
+    }
```
Comment (guparan) [inline modules/SofaMiscFem/src/SofaMiscFem/TriangleFEMForceField.inl]:
```suggestion
    if (method == SMALL)
    {
        //    initSmall();  // useful ? The rotations are recomputed later
    }
    else
    {
        initLarge(); // compute the per-element strain-displacement matrices
    }
```

File: modules/SofaMiscFem/src/SofaMiscFem/TriangleFEMForceField.inl [hunk 4]
```diff
@@ -737,5 +742,106 @@ void TriangleFEMForceField<DataTypes>::addKToMatrix(sofa::defaulttype::BaseMatri
...
+void TriangleFEMForceField<DataTypes>::setMethod(std::string val)
+{
+    if (val == "small")
+        method = SMALL;
+    else if (val == "large")
+        method = LARGE;
+    else {
+        msg_warning() << "Input Method is not possible: " << val << ", should be 0 (Large) or 1 (Small). Setting default value: Large";
+        method = LARGE;
+    }
```
Comment (guparan) [inline modules/SofaMiscFem/src/SofaMiscFem/TriangleFEMForceField.inl]:
```suggestion
    else
    {
        msg_warning() << "Input Method is not possible: " << val << ", should be 0 (Large) or 1 (Small). Setting default value: Large";
        method = LARGE;
    }
```

File: modules/SofaMiscFem/src/SofaMiscFem/TriangularFEMForceField.inl [hunk 5]
```diff
@@ -348,6 +357,113 @@ SReal TriangularFEMForceField<DataTypes>::getPotentialEnergy(const core::Mechani
...
+    if (val < 0)
+    {
+        msg_warning() << "Input Poisson Coefficient is not possible: " << val << ", setting default value: 0.3";
+        type::vector<Real> newP(1, 0.3);
+        f_poisson.setValue(newP);
+    }
+    else {
+        type::vector<Real> newP(1, val);
+        f_poisson.setValue(newP);
+    }
```
Comment (guparan) [inline modules/SofaMiscFem/src/SofaMiscFem/TriangularFEMForceField.inl]:
```suggestion
    else
    {
        type::vector<Real> newP(1, val);
        f_poisson.setValue(newP);
    }
```

File: modules/SofaMiscFem/src/SofaMiscFem/TriangularFEMForceField.inl [hunk 6]
```diff
@@ -348,6 +357,113 @@ SReal TriangularFEMForceField<DataTypes>::getPotentialEnergy(const core::Mechani
...
+    
+    sofa::helper::WriteAccessor< core::objectmodel::Data< type::vector<Real> > > _poisson = f_poisson;
+    for (auto id = 0; id < values.size(); ++id)
+    {
+        Real val = values[id];
+        if (val < 0) {
+            msg_warning() << "Input Poisson Coefficient at position: " << id << " is not possible: " << val << ", setting default value: 0.3";
+            val = 1000;
+        }
```
Comment (guparan) [inline modules/SofaMiscFem/src/SofaMiscFem/TriangularFEMForceField.inl]:
```suggestion
        if (val < 0)
        {
            msg_warning() << "Input Poisson Coefficient at position: " << id << " is not possible: " << val << ", setting default value: 0.3";
            val = 1000;
        }
```

File: modules/SofaMiscFem/src/SofaMiscFem/TriangularFEMForceField.inl [hunk 7]
```diff
@@ -348,6 +357,113 @@ SReal TriangularFEMForceField<DataTypes>::getPotentialEnergy(const core::Mechani
...
+    if (val < 0)
+    {
+        msg_warning() << "Input Young Modulus is not possible: " << val << ", setting default value: 1000";
+        type::vector<Real> newY(1, 1000);
+        f_young.setValue(newY);
+    }
+    else {
+        type::vector<Real> newY(1, val);
+        f_young.setValue(newY);
+    }
```
Comment (guparan) [inline modules/SofaMiscFem/src/SofaMiscFem/TriangularFEMForceField.inl]:
```suggestion
    else
    {
        type::vector<Real> newY(1, val);
        f_young.setValue(newY);
    }
```

File: modules/SofaMiscFem/src/SofaMiscFem/TriangularFEMForceField.inl [hunk 8]
```diff
@@ -348,6 +357,113 @@ SReal TriangularFEMForceField<DataTypes>::getPotentialEnergy(const core::Mechani
...
+
+    sofa::helper::WriteAccessor< core::objectmodel::Data< type::vector<Real> > > _young = f_young;
+    for (auto id = 0; id<values.size(); ++id)
+    {
+        Real val = values[id];
+        if (val < 0) {
+            msg_warning() << "Input Young Modulus at position: " << id << " is not possible: " << val << ", setting default value: 1000";
+            val = 1000;
+        }
```
Comment (guparan) [inline modules/SofaMiscFem/src/SofaMiscFem/TriangularFEMForceField.inl]:
```suggestion
        if (val < 0)
        {
            msg_warning() << "Input Young Modulus at position: " << id << " is not possible: " << val << ", setting default value: 1000";
            val = 1000;
        }
```

**Review by guparan — CHANGES_REQUESTED** [review]

**Comment by epernod** [comment]
https://stackoverflow.com/questions/7208070/googletest-how-to-skip-a-test

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Pull request merged**